### ISSUE TITLE
Removal of CPU ID checking

### ISSUE THREAD

**Issue by rogerclarkmelbourne** [rogerclarkmelbourne/Radioddity_GD-77#21]
Both the bootloader and the firmware check some data at 0x7f800 in an attempt to prevent the firmware binary from being copied

Kai has already reverse engineered the algorithm using to generate the "key" data, and we can't do anything about the bootloader checking the key. 

But to save problems in the future we should establish where the CPU ID checking code is located in the firmware and disable it.

I did take a quick look at this the other night, but I didnt quite work out how to disable it (as I was too tired ;-))

Address 0x7800 is used in the function at 0x65008, which seems to simply fetch the bytes

There are 2 functions which seem to use the key bytes that have been fetched, one of them is
FUN_00065290

At some point the code must compare the decoded key bytes with the CPU ID, retrieved by
CPU_get_unique_id_00055820

And looking at the bottom of this function, it calls  FUN_0005b84a, so I think perhaps FUN_0005b84a is the validation function.

I think I'll try patching this and seeing what happens ;-)

**Comment by rogerclarkmelbourne** [comment]
umm
CPU_get_unique_id_00055820 doesnt seem to get called during normal bootup

**Comment by forkoz** [comment]
sub_6A1BC which is called from the main task has a statement if ( sub_6511A() )

6511A gets CPUID multiple ways, sets some dwords in ram and returns different results

**Comment by forkoz** [comment]
In the bootloader its a very similar function: sub_1D28


    MK22_Pin_CTRL_Reg_n_PORTB_PCR18 = 0x100;
    MK22_Port_Data_Direction_Reg_GPIOB_PDDR[0] |= 0x40000u;
    MK22_Port_Set_Output_Reg_GPIOB_PSOR[0] |= 0x40000u;
    CPUID_bl_maybe_1CF6((int)&v9);
    for ( i = 0; i < 4; ++i )
      *(&v9 + i) = 0x1Du / (*(&v9 + i) + 1) - *(&v9 + i) - 1;
    chex_cpuid_flash(16, (int)&v10, 0x10u);
    for ( j = 0; j < 4; j = (unsigned __int8)(j + 1) )

also called as a condition:
 if ( sub_1D28() )
inside sub_3580

**Comment by forkoz** [comment]
Or just edit the function that gets the flash ID to retrieve the HW register instead.

**Comment by forkoz** [comment]
CPUID_CHECK_2_653B8 is another good contender in the main task. 

I think it's stored in the memory:
1FFF1990 MatchCPUID_1FFF1990 DCB 1 << result of checking
1FFF1992 CPUID_something_1FFF1992 DCW 0x64 << CPUID

**Comment by forkoz** [comment]
Heh, probably if you patch getCPUID_dword_6522C to always return 1 it will stop checking in the main and menu tasks. I don't think any other place will check except sub_6A1BC

**Comment by rogerclarkmelbourne** [comment]
OK
I'll try that

Actually the easiest way to test is to just change the result to false in the debugger

**Comment by forkoz** [comment]
True...  I think I boned that first patch because it needs the function to execute the stuff it would have if CPUid check was true. The 2nd one I patched and it decompiles corectly.

**Comment by rogerclarkmelbourne** [comment]
It seems to check multiple times at startup.

If I set $r0=0 ever times it breaks, I eventually get the "no activation" message

**Comment by rogerclarkmelbourne** [comment]
This should do it

                             **************************************************************
                             *                          FUNCTION                          *
                             **************************************************************
                             undefined FUN_0006522c()
             undefined         r0:1           <RETURN>
                             FUN_0006522c                                    XREF[5]:     FUN_00065066:00065068(c), 
                                                                                          0006533c(c), 0006535a(c), 
                                                                                          FUN_00065518:0006551e(c), 
                                                                                          Task_0x01_main:0006a18a(c)  
        0006522c 00 bf           nop
        0006522e 01 20           mov        r0,#0x1
        00065230 70 47           bx         lr

**Comment by forkoz** [comment]
yep,  use CMP r0, r0 as nop.

ErrorCheck_cpuidresult_652F0 is where those errors come from.

**Comment by rogerclarkmelbourne** [comment]
I was also using mov r0,r0 as nop
but nop also seems OK
perhaps nop is actually slower

**Comment by forkoz** [comment]
dunno, its just habit.

**Comment by forkoz** [comment]
In main task though it will still check branching from sub_6A1BC();

 lpm_set_clock_configuration(0);
  sub_6A1BC();
  sub_201DA();
  if ( getCPUID_dword_6522C() == 1 )
    sub_1261E();
  else
    CPUID_doublecheck_65532();

and 
int sub_6A1BC()
{
  sub_6A1E2();
  sub_6A1F2();
  if ( cpuidCheck_6511A() )
  {
    sub_6A20C();
    sub_6A27E();
    sub_12772();
  }
  return mouse_test();
}

that sub does a real check, not just dword.

**Comment by rogerclarkmelbourne** [comment]
yes
just patching that return value didn't fix it

**Comment by forkoz** [comment]
ROM:0006A1BC sub_6A1BC                               ; CODE XREF: Task_0x01_main+8p
ROM:0006A1BC                 PUSH    {R7,LR}         ; Push registers
ROM:0006A1BE                 BL      sub_6A1E2       ; Branch with Link
ROM:0006A1C2                 BL      sub_6A1F2       ; Branch with Link
ROM:0006A1C6                 BL      cpuidCheck_6511A ; Branch with Link  << this needs to go
ROM:0006A1CA                 CMP     R0, #0          ; Set cond. codes on Op1 - Op2
ROM:0006A1CC                 BEQ     loc_6A1DA       ; Branch << before it leaves
ROM:0006A1CE                 BL      sub_6A20C       ; Branch with Link << but these need to run
ROM:0006A1D2                 BL      sub_6A27E       ; Branch with Link <<
ROM:0006A1D6                 BL      sub_12772       ; Branch with Link  <<
ROM:0006A1DA
ROM:0006A1DA loc_6A1DA                               ; CODE XREF: sub_6A1BC+10j
ROM:0006A1DA                 POP.W   {R0,LR}         ; Pop registers
ROM:0006A1DE                 B.W     mouse_test      ; Branch

**Comment by forkoz** [comment]
I just did this: 
0006A1BC sub_6A1BC                               ; CODE XREF: Task_0x01_main+8p
ROM:0006A1BC                 PUSH    {R7,LR}         ; Push registers
ROM:0006A1BE                 BL      sub_6A1E2       ; Branch with Link
ROM:0006A1C2                 BL      sub_6A1F2       ; Branch with Link
ROM:0006A1C6                 NOP                     ; No Operation
ROM:0006A1C8                 NOP                     ; No Operation
ROM:0006A1CA                 NOP                     ; No Operation
ROM:0006A1CC                 NOP                     ; No Operation
ROM:0006A1CE                 BL      sub_6A20C       ; Branch with Link
ROM:0006A1D2                 BL      sub_6A27E       ; Branch with Link
ROM:0006A1D6                 BL      sub_12772       ; Branch with Link
ROM:0006A1DA                 POP.W   {R0,LR}         ; Pop registers
ROM:0006A1DE                 B.W     mouse_test      ; Branch

**Comment by rogerclarkmelbourne** [comment]
Did that work ? Or do you want me to try that

**Comment by forkoz** [comment]
Try it, if you can

**Comment by rogerclarkmelbourne** [comment]
OK

**Comment by forkoz** [comment]
This is how it plays out:

void __noreturn Task_0x01_main()
{
  int *v0; // r0@5
  int v1; // r1@5
  int v2; // r2@5
  int v3; // r3@5
  int v4; // r1@5
  __int64 v5; // r2@5

  lpm_set_clock_configuration(0);
  sub_6A1BC();                                  // //fixed by nops I hope
  sub_201DA();
  if ( getCPUID_dword_6522C() == 1 )  //our dword grabber
    sub_1261E();                                // returns startup key if held
  else
    CPUID_doublecheck_65532();                  // serious cpu check
  int_install_unexpected_isr();
  sub_1333C(0);
  while ( 1 )
  {
    v0 = sub_132B6();
    sub_201F4((int)v0, v1, v2, v3);
    time_delay_ticks(1u, v4, v5);
    sub_13354(0);
  }
}

**Comment by rogerclarkmelbourne** [comment]
I manually patched it by typing in 00 bf 00 bf 00 bf 00 bf and I think its working

`
                             **************************************************************
                             *                          FUNCTION                          *
                             **************************************************************
                             void __stdcall Startup_0006a1bc(undefined4 uParm1, undef
             void              <VOID>         <RETURN>
             undefined4        r0:4           uParm1
             undefined4        r1:4           uParm2
             undefined4        r2:4           uParm3
             undefined4        r3:4           uParm4
                             Startup_0006a1bc                                XREF[1]:     Task_0x01_main:0006a182(c)  
        0006a1bc 80 b5           push       { r7, lr }
        0006a1be 00 f0 10 f8     bl         FUN_0006a1e2                                     undefined FUN_0006a1e2()
        0006a1c2 00 f0 16 f8     bl         FUN_0006a1f2                                     undefined FUN_0006a1f2()
        0006a1c6 00 bf           nop
        0006a1c8 00 bf           nop
        0006a1ca 00 bf           nop
        0006a1cc 00 bf           nop
        0006a1ce 00 f0 1d f8     bl         Startup_0006a20c                                 void Startup_0006a20c(undefined4
        0006a1d2 00 f0 54 f8     bl         Startup_0006a27e                                 void Startup_0006a27e(void)
        0006a1d6 a8 f7 cc fa     bl         FUN_00012772                                     undefined FUN_00012772()
        0006a1da bd e8 01 40     pop.w      {uParm1, lr }
        0006a1de d2 f7 43 bc     b.w        FUN_0003ca68                                     undefined FUN_0003ca68()
                             -- Flow Override: CALL_RETURN (CALL_TERMINATOR)

`

**Comment by forkoz** [comment]
Yea, that's what I did. Its cool that works in the live debugger. Does it ever pull the hardware CPUid anymore?

Now if we could only get the boot loader.... well you can if you flash it with HW tools.

**Comment by rogerclarkmelbourne** [comment]
I put a read watch on 40048054 but it doesnt trigger, but I'm not sure that I can watch the hardware registers memory mapped registers using GDB

I tried doing this before and didnt have much luck.

**Comment by rogerclarkmelbourne** [comment]
Patching the bootloader is risky.

**Comment by forkoz** [comment]
True, it only works for people that can HW flash... once you have a tested method you can make a special firmware that flashes from those compiled functions you inject.

**Comment by rogerclarkmelbourne** [comment]
Yes.
The main application could reflash the bootloader, but its still risky, because if it goes wrong the radio is bricked until you connect via SWD

Normally its hard to damage the bootloader even if the main app crashes

**Comment by forkoz** [comment]
Didn't the MD380 guys take the encryption off? This all only comes into play if we need the space in flash or we need to share backups.

**Comment by rogerclarkmelbourne** [comment]
AFIK

The when Travis first kinda accidentally downloaded the bootloader, when he was writing a codeplug utility, he patched the bootloader code to stop it read protecting the MCU.

I think perhaps Kai did the same thing, because the bootloader I am using is from a full dump of the ROM that Kai sent me, and he may have already patched the bootloader and reflashed it.

Unfortunately because the bootloader that will be in everyone else’s radios will still check the key at 0x7f800, which means we can’t use absolutely all the ROM, because it would overwrite the bootloader key values.

I just thought it was safer to remove the CPU ID checking code in the main application, as it’s one less thing to go wrong.
Also we may be able to reuse the ROM space taken by those functions, potentially for some small functions that we create

**Comment by forkoz** [comment]
I don't think he patched it since he made the calculator. It looks trivial to fix the check, the non trivial part as you say is safely uploading the modified bootloader. CPUID stuff looked pretty big, like you could fit a good sized function in there. It also takes away the other error modes in the way that it's patched. eg "no app software" and "low bt version" for what that's worth

we made contiguous space at least from 0x653B8 - 655B0, conservatively.

**Comment by KG5RKI** [comment]
There was a bug in the USB protocol.  Off the top of my head right now, but basically it sends some initial commands, then sends a command to set the block number, then sends one telling it to read and so on.. basically the check on the actual address to be read is done in the set address command, and if the address is not in an allowed range it returns.  SO what happens if you don't send the set address command and then immediately just do a read block command? the pointer for address to be read defaults to an address that points to the bootloader, so you just read it out, then reverse the code for encryption and hope its not tied to something console specific xD  unfortunately this was not the case for md2017 / RT82, but they made another pretty bad mistake of just using a static repeating 512byte xor block with no shift, so it was easy to figure it out.

**Comment by rogerclarkmelbourne** [comment]
This is a bug in the MD-380. The GD-77 and DMR-1801 etc don't use the same system at all

**Comment by KG5RKI** [comment]
yea I know, I've been studying your code and understanding the history and steps ya'll took to get this put together.  Extremely impressive, and so much potential with this.  Props to yall.  I am so surprised at how far along you already are with this and how well it works.  (been using a fork off your branch roger for tier2).  Really great stuff

**Comment by rogerclarkmelbourne** [comment]
The Tier2 branch is extremely unstable.

I keep needing to revert it when I make a change too far.

I think I'll need to change my backup process and not keep pushing to git, as often I find after I've pushed to git, that something I'd not checked is broken

I think the last few commits no longer work for simplex and only work for Rx from a repeater.

**Comment by KG5RKI** [comment]
Its all good man, and yea its unstable but it helps when I can see your process of debugging.  I'm more interested in understanding the crucial parts of getting radios of this type (hardware IC combos) (HD1, anytone, etc..) to work on tier II and etc.. We should talk sometime and bounce ideas around.  I'm still studying and catching up but I've got it compiling and flashing and I'll make a pull request soon for the updated scripts for linux.  Tweaked your python example to allow flashing straight .bin from the IDE and also supports the container format, and also made a script for bin2gpl so you dont have to use mono and the current exe converter.. I run linux so this was a must have for me and easily implemented by following the flashtool src.  Now I can start playing :D Thank you for the new toy

**Comment by KG5RKI** [comment]
do you guys have a discord or telegram room or something?

**Comment by rogerclarkmelbourne** [comment]
I am the only dev on the project.

Kai, DG4LKU, who started the project, withdrew just over 2 months ago.

No. There is no forum etc. I'm not on telegram etc. I'm not on social media either.
